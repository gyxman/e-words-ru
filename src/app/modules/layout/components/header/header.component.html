<div class="info">
    <button
        automation-id="toggle"
        mat-icon-button
        color="primary"
        aria-label="Открыть меню"
    >
        <mat-icon
            svgIcon="toggle"
            aria-hidden="false"
            aria-label="Иконка открытия меню"
        ></mat-icon>
    </button>
    <a class="logo" automation-id="logo" [routerLink]="['/']">Тренажер слов</a>
</div>
<div class="actions">
    <button
        automation-id="exit"
        mat-icon-button
        color="primary"
        aria-label="Выйти из приложения"
    >
        <mat-icon
            svgIcon="logout"
            aria-hidden="false"
            aria-label="Иконка выхода из приложения"
        ></mat-icon>
    </button>
</div>
