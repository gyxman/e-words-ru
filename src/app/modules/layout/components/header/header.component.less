@import '../../../../../styles/index';

:host {
    display: flex;
    align-items: center;
    justify-content: space-between;
    padding: 12px 16px;
    background-color: @color-gray-extra-light2;
}

.info {
    display: flex;
    align-items: center;
}

.logo {
    .font-h4();
    margin-left: 16px;
    text-decoration: none;
    color: unset;
}
