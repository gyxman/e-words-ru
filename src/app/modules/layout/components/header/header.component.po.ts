import {PageObject} from '../../../../testing/page-object';
import {DebugElement} from '@angular/core';

export class HeaderComponentPo<T> extends PageObject<T> {
    get openMenu(): DebugElement {
        return this.getByAutomationId('open-menu');
    }

    get exit(): DebugElement {
        return this.getByAutomationId('exit');
    }
}
