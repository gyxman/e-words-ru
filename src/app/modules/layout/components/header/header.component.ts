import {Component} from '@angular/core';
import {MatIconRegistry} from '@angular/material/icon';
import {DomSanitizer} from '@angular/platform-browser';

@Component({
    selector: 'app-header',
    templateUrl: './header.component.html',
    styleUrls: ['./header.component.less'],
})
export class HeaderComponent {
    constructor(iconRegistry: MatIconRegistry, sanitizer: DomSanitizer) {
        iconRegistry.addSvgIcon(
            'logout',
            sanitizer.bypassSecurityTrustResourceUrl('assets/icons/icon-logout.svg'),
        );
        iconRegistry.addSvgIcon(
            'toggle',
            sanitizer.bypassSecurityTrustResourceUrl('assets/icons/icon-toggle.svg'),
        );
    }
}
