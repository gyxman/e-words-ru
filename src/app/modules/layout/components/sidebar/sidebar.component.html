<ul class="list">
    <li
        *ngFor="let item of allMenuItems$ | async; trackBy: identityTrackByFunction"
        class="item"
    >
        <a class="link" [routerLink]="item.navigateTo">{{ item.title }}</a>
    </li>
</ul>
