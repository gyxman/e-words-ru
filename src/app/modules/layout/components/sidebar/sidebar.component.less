@import '../../../../../styles/index';

:host {
    display: block;
}

.list {
    padding: 10px 0;
    margin: 0;
    list-style: none;
}

.link {
    .font-h5();
    display: block;
    padding: 10px 20px;
    text-decoration: none;
    color: inherit;

    &:hover {
        background-color: @color-gray-extra-light;
    }
}
