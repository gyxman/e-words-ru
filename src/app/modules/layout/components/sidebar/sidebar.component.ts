import {Component, EventEmitter, HostListener, Output} from '@angular/core';
import {SidebarFacadeService} from '../../services/sidebar-facade.service';
import {SidebarMenuItem} from '../../models/sidebar-menu-item';

@Component({
    selector: 'app-sidebar',
    templateUrl: './sidebar.component.html',
    styleUrls: ['./sidebar.component.less'],
})
export class SidebarComponent {
    @Output() sidebarOnClick = new EventEmitter();

    @HostListener('click') onClick() {
        this.sidebarOnClick.emit();
    }

    readonly allMenuItems: SidebarMenuItem[] = this.sidebarFacadeService.allMenu;

    constructor(private sidebarFacadeService: SidebarFacadeService) {}
}
