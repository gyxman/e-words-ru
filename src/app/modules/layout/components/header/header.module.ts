import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {HeaderComponent} from './header.component';
import {MatButtonModule} from '@angular/material/button';
import {MatIconModule} from '@angular/material/icon';
import {RouterModule} from '@angular/router';

@NgModule({
    imports: [CommonModule, MatButtonModule, MatIconModule, RouterModule],
    declarations: [HeaderComponent],
    exports: [HeaderComponent],
})
export class HeaderModule {}
