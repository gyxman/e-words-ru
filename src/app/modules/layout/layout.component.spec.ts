import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {LayoutComponent} from './layout.component';
import {RouterTestingModule} from '@angular/router/testing';
import {MockComponent} from 'ng-mocks';
import {HeaderComponent} from './components/header/header.component';
import {SidebarComponent} from './components/sidebar/sidebar.component';
import {MatSidenavModule} from '@angular/material/sidenav';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import {LayoutComponentPo} from './layout.component.po';
import {LayoutFacadeService} from './services/layout-facade.service';
import {instance, mock, when} from 'ts-mockito';
import {BehaviorSubject} from 'rxjs';
import {LoaderModule} from '../utils/components/loader/loader.module';

describe('LayoutComponent - компонент авторизованной зоны', () => {
    let component: LayoutComponent;
    let fixture: ComponentFixture<LayoutComponent>;
    let pageObject: LayoutComponentPo<LayoutComponent>;
    let layoutFacadeServiceMock: LayoutFacadeService;
    let showLoaderMock$: BehaviorSubject<boolean>;

    beforeEach(() => {
        layoutFacadeServiceMock = mock(LayoutFacadeService);

        showLoaderMock$ = new BehaviorSubject(false);
    });

    beforeEach(() => {
        when(layoutFacadeServiceMock.showLoader$).thenReturn(showLoaderMock$);
    });

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            declarations: [
                LayoutComponent,
                MockComponent(HeaderComponent),
                MockComponent(SidebarComponent),
            ],
            imports: [
                RouterTestingModule,
                MatSidenavModule,
                NoopAnimationsModule,
                LoaderModule,
            ],
            providers: [
                {
                    provide: LayoutFacadeService,
                    useFactory: () => instance(layoutFacadeServiceMock),
                },
            ],
        }).compileComponents();
    }));

    beforeEach(() => {
        fixture = TestBed.createComponent(LayoutComponent);
        component = fixture.componentInstance;
        pageObject = new LayoutComponentPo(fixture);
    });

    describe('Проверка общей верстки компонента', () => {
        it('Если компонент загрузился, то все элементы отрисовались', () => {
            // act
            fixture.detectChanges();

            // assert
            expect(fixture.nativeElement).toMatchSnapshot();
        });
    });

    describe('Отображение лоадера', () => {
        it('Если информация о показе лоадера не пришла, то не отображаем его на странице', () => {
            // arrange
            showLoaderMock$.next(false);

            // act
            fixture.detectChanges();

            // assert
            expect(pageObject.loader.showLoader).toBe(false);
        });

        it('Если информация о показе лоадера пришла, то отображаем его на странице', () => {
            // arrange
            showLoaderMock$.next(true);

            // act
            fixture.detectChanges();

            // assert
            expect(pageObject.loader.showLoader).toBe(true);
        });

        it('Если информация о показе лоадера пришла, то у лоадера есть оверлей', () => {
            // arrange
            showLoaderMock$.next(true);

            // act
            fixture.detectChanges();

            // assert
            expect(pageObject.loader.showOverlay).toBe(true);
        });
    });

    describe('Левая панель', () => {
        it('Если компонент загрузился, то левая панель скрыта', () => {
            // act
            fixture.detectChanges();

            // assert
            expect(pageObject.sidebarContent).toBeNull();
        });

        it('Если из header пришло событие открытия сайдбара, то левая панель открывается', () => {
            // arrange
            fixture.detectChanges();

            // act
            pageObject.header.triggerEventHandler('openSidebar', null);

            fixture.detectChanges();

            // assert
            expect(pageObject.sidebarContent).toBeTruthy();
        });

        it('Если пользователь открывает левую панель, а потом нажимает esc, то левая панель закрывается', () => {
            // arrange
            fixture.detectChanges();

            pageObject.header.triggerEventHandler('openSidebar', null);

            fixture.detectChanges();

            // act
            pageObject.sidebar.triggerEventHandler('keydown.escape', null);

            fixture.detectChanges();

            // assert
            expect(pageObject.sidebarContent).toBeNull();
        });

        it(`Если пользователь открывает левую панель, а потом нажимает на подложку в основной части приложения,
        то левая панель закрывается`, () => {
            // arrange
            fixture.detectChanges();

            pageObject.header.triggerEventHandler('openSidebar', null);

            fixture.detectChanges();

            // act
            pageObject.layout.triggerEventHandler('backdropClick', null);

            fixture.detectChanges();

            // assert
            expect(pageObject.sidebarContent).toBeNull();
        });
    });
});
