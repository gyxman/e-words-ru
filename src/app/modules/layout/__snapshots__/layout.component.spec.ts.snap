// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutComponent - компонент авторизованной зоны Проверка общей верстки компонента Если компонент загрузился, то все элементы отрисовались 1`] = `
<div
  id="root0"
>
  <app-loader
    automation-id="loader"
  >
    <div
      automation-id="content"
      class="content"
    >
      <mat-sidenav-container
        automation-id="layout"
        class="layout mat-drawer-container mat-sidenav-container"
      >
        
        <div
          class="mat-drawer-backdrop"
        />
        <div
          aria-hidden="true"
          class="cdk-visually-hidden cdk-focus-trap-anchor"
        />
        <mat-sidenav
          automation-id="sidebar"
          class="sidebar mat-drawer mat-sidenav ng-trigger ng-trigger-transform mat-drawer-over"
          style="box-shadow:none;visibility:hidden;0:box-shadow;1:visibility;"
          tabindex="-1"
        >
          <div
            class="mat-drawer-inner-container"
          >
            
          </div>
        </mat-sidenav>
        <div
          aria-hidden="true"
          class="cdk-visually-hidden cdk-focus-trap-anchor"
        />
        <mat-sidenav-content
          class="wrapper mat-drawer-content mat-sidenav-content"
        >
          <app-header
            automation-id="header"
            class="header"
          />
          <div
            class="content"
          >
            <div
              class="container"
            >
              <router-outlet />
            </div>
          </div>
        </mat-sidenav-content>
        
      </mat-sidenav-container>
    </div>
    
  </app-loader>
</div>
`;
